src/librustc_interface/interface.rs

```diff
@@ -163,40 +163,50 @@ pub fn run_compiler_in_existing_thread_pool<R>(
     config: Config,
     f: impl FnOnce(&Compiler) -> R,
 ) -> R {
-    let registry = &config.registry;
     let (sess, codegen_backend) = util::create_session(
         config.opts,
         config.crate_cfg,
         config.diagnostic_output,
         config.file_loader,
         config.input_path.clone(),
         config.lint_caps,
-        registry.clone(),
+        config.registry.clone(),
     );
 
-    let compiler = Compiler {
-        sess,
-        codegen_backend,
-        input: config.input,
-        input_path: config.input_path,
-        output_dir: config.output_dir,
-        output_file: config.output_file,
-        crate_name: config.crate_name,
-        register_lints: config.register_lints,
-        override_queries: config.override_queries,
-    };
-
-    let r = {
-        let _sess_abort_error = OnDrop(|| {
-            compiler.sess.finish_diagnostics(registry);
-        });
-
-        f(&compiler)
-    };
-
-    let prof = compiler.sess.prof.clone();
-    prof.generic_activity("drop_compiler").run(move || drop(compiler));
-    r
+    let registry = &config.registry;
+    let input = config.input;
+    let input_path = config.input_path;
+    let output_dir = config.output_dir;
+    let output_file = config.output_file;
+    let crate_name = config.crate_name;
+    let register_lints = config.register_lints;
+    let override_queries = config.override_queries;
```

Comment:
Nit: you can move the whole `compiler` into the closure - https://github.com/petrochenkov/rust/commit/17d53ecba9da4a2ced169ca040e902ff2edbc293.